src/snowflake/connector/network_async.py

```diff
@@ -0,0 +1,465 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import asyncio
+
+import aiohttp
+
+from .network import *
+
+# YICHUAN: If we don't want to duplicate these raise  functions we can also modify them to directly take the error
+# status and reason instead of a response object (that way we can use the same for both sync and async)
+
+
+# For this commit I'm just adding separate functions for async so that I don't need to change too much code
+def raise_okta_unauthorized_error_async(
+    connection: SnowflakeConnection | None, response: aiohttp.ClientResponse
+) -> None:
+    Error.errorhandler_wrapper(
+        connection,
+        None,
+        DatabaseError,
+        {
+            "msg": f"Failed to get authentication by OKTA: {response.status}: {response.reason}",
+            "errno": ER_FAILED_TO_CONNECT_TO_DB,
+            "sqlstate": SQLSTATE_CONNECTION_REJECTED,
+        },
+    )
+
+
+def raise_failed_request_error_async(
+    connection: SnowflakeConnection | None,
+    url: str,
+    method: str,
+    response: aiohttp.ClientResponse,
+) -> None:
+    # YICHUAN: NO TELEMETRY FOR NOW
+
+    # TelemetryService.get_instance().log_http_request_error(
+    #     f"HttpError{response.status}",
+    #     url,
+    #     method,
+    #     SQLSTATE_CONNECTION_WAS_NOT_ESTABLISHED,
+    #     ER_FAILED_TO_REQUEST,
+    #     response=response,
+    # )
+    Error.errorhandler_wrapper(
+        connection,
+        None,
+        InterfaceError,
+        {
+            "msg": f"{response.status} {response.reason}: {method} {url}",
+            "errno": ER_FAILED_TO_REQUEST,
+            "sqlstate": SQLSTATE_CONNECTION_WAS_NOT_ESTABLISHED,
+        },
+    )
+
+
+# Yichuan: CURRENTLY UNUSED, see SnowflakeRestfulAsync._request_exec_async
+class SnowflakeAuthAsync(aiohttp.BasicAuth):
```

Comment:
this is on par with `SnowflakeAuth` class.
in the current implementation we directly update the header dict